packages/data/src/components/use-select/index.js

```diff
@@ -207,7 +207,7 @@ export default function useSelect( mapSelect, deps ) {
 	const isMounted = useRef( false );
 
 	useIsomorphicLayoutEffect( () => {
-		if ( ! hasMappingFunction ) {
+		if ( ! hasMappingFunction || isAsync === null ) {
```

Comment:
Either we create a new possible value `null` here, meaning "don't subscribe and update", or we need a separate context provider.

Reply:
So, in addition to an "async" mode, we now also have a "nosync" mode 🙂 it's turned on when mounting the provider as `<AsyncModeProvider value={ null } />` which is not particularly intuitive but anyway.

One unwanted consequence of having `isAsync` as a dependency of this effect is that it will run (and resubscribe) also when the value changes between `false` and `true`. We don't want that, the subscription should stay the same. We only want to react to changes between `null` and `true | false`. Let's do something like:
```js
const isSyncing = isAsync != null;
useEffect( () => {
  if ( ! isSyncing ) {
    return;
  }
  // subscription mgmt stuff
}, [ isSyncing, /* ... */ ] );
```

Another place that needs updating is the `hasLeftAsyncMode` condition a few lines above. Newly we want to force running the `mapSelect` function also when `isAsync` changes from `null` to `true | false`. Otherwise it will render the stale `latestMapOutput` value that was computed either on first render or just before the component went to nosync mode.

We want to ensure (and also unit test) that merely rerendering `AsyncModeProvider` with a changed `value` prop will cause the children to read the latest data store state. In current Gutenberg the `value` change is probably always associated with data store state change (I'd guess `blockVisibility`?), so the observable UI behavior is good, but under different circumstances it could break.